src/mlpack/methods/ann/rbm_impl.hpp

```diff
@@ -0,0 +1,237 @@
+
+  if (!reset)
+    Reset();
+
+  // Train the model.
+  Timer::Start("rbm_optimization");
+  optimizer.Optimize(*this, parameter);
+  Timer::Stop("rbm_optimization");
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+double RBM<InitializationRuleType, RBMPolicy>::
+    FreeEnergy(arma::Mat<ElemType>&& input)
+{
+  return rbmPolicy.FreeEnergy(std::move(input));
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+double RBM<InitializationRuleType, RBMPolicy>::Evaluate(
+    const arma::Mat<ElemType>& /* parameters*/,
+    const size_t i,
+    const size_t /*batchSize*/)
+{
+  if (!useMonitoringCost)
+  {
+    Gibbs(std::move(predictors.col(i)), std::move(negativeSamples));
+    return std::fabs(FreeEnergy(std::move(predictors.col(i))) -
+        FreeEnergy(std::move(negativeSamples)));
+  }
+  else
+  {
+    if (persistence)
+    {
+      return rbmPolicy.Evaluate(predictors, i, predictors.n_cols);
+    }
+    else
+    {
+      // Mean Squared Error
+      rbmPolicy.SampleHidden(std::move(predictors.col(i)),
+          std::move(hiddenReconstruction));
+      rbmPolicy.SampleVisible(std::move(hiddenReconstruction),
+          std::move(visibleReconstruction));
+      return arma::accu(arma::pow(visibleReconstruction - predictors.col(i),
+          2)) / hiddenReconstruction.n_rows;
+    }
+  }
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::SampleHidden(
+    arma::Mat<ElemType>&& input,
+    arma::Mat<ElemType>&& output)
+{
+  rbmPolicy.SampleHidden(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::SampleVisible(
+    arma::Mat<ElemType>&& input,
+    arma::Mat<ElemType>&& output)
+{
+  rbmPolicy.SampleVisible(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::Gibbs(
+    arma::Mat<ElemType>&& input,
+    arma::Mat<ElemType>&& output,
+    size_t steps)
+{
+  steps = (steps == SIZE_MAX) ? this-> numSteps: steps;
+
+  if (persistence && !state.is_empty())
+  {
+    SampleHidden(std::move(state), std::move(gibbsTemporary));
+    SampleVisible(std::move(gibbsTemporary), std::move(output));
+  }
+  else
+  {
+    SampleHidden(std::move(input), std::move(gibbsTemporary));
+    SampleVisible(std::move(gibbsTemporary), std::move(output));
+  }
+
+  for (size_t j = 1; j < steps; j++)
+  {
+    SampleHidden(std::move(output), std::move(gibbsTemporary));
+    SampleVisible(std::move(gibbsTemporary), std::move(output));
+  }
+  if (persistence)
+    state = output;
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::Gradient(
+    arma::Mat<ElemType>& /*parameters*/,
+    const size_t input,
+    arma::Mat<ElemType>& output,
+    const size_t /*batchSize*/)
+{
+  positiveGradient.zeros();
```

Comment:
Sorry, I didn't see your comment. You should repeat the procedure below for each feature from `input` to `input + batchSize - 1`. The present implementation omits all features except the first one. That's why the accuracy is wrong.